src/app.h

```diff
@@ -0,0 +1,384 @@
+#ifndef APP_H
+#define APP_H
+
+#include "logging.h"
+#include "model.h"
+#include "gui.h"
+
+#include <imgui.h>
+#include <opencv2/core.hpp>
+
+#include <map>
+#include <queue>
+
+namespace var {
+
+using MinPriorQueue = std::priority_queue<int, std::vector<int>, std::greater<int>>;
+
+class app {
+public:
+    app():
+        _gui("ELTEVar", { 1280, 720 }),
+        _model("res/pitch.png")
+    {
+        glGenTextures(1, &_state.pitchTexture);
+        glBindTexture(GL_TEXTURE_2D, _state.pitchTexture);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+
+        // Set texture clamping method
+        const auto& pitch = _model.pitch();
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP);
+        glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, pitch.cols, pitch.rows, 0, GL_RGBA, GL_UNSIGNED_BYTE, pitch.ptr());
+    }
+
+    void start() {
+        _gui.run([this] {
+            _lastUpdate = std::chrono::steady_clock().now().time_since_epoch();
+            _state.width = _gui.width();
+            _state.height = _gui.height();
+
+            renderMainMenu();
+            renderWindows();
+
+            _delta = std::chrono::steady_clock().now().time_since_epoch() - _lastUpdate;
+        });
+    }
+
+    void renderMainMenu() {
+        _gui.main_menu().menu("File")
+            .item("Load", [this] {
+                _state.loadPopupFlag = true;
+
+                if (_state.debugWindowFlag) {
+                    _model.imagePath() = "/Users/lafette21/Downloads/video-assisted-referee/data/DSCF0137.jpeg";
+                }
+            })
+            .item("Save", [this] { _state.savePopupFlag = true; });
+
+        _gui.main_menu().menu("Edit")
+            .item("Generate", [this] {
+                try {
+                    auto [image, pitch] = _model.generate();
+                    cv::cvtColor(image, image, cv::COLOR_RGBA2BGR);
+                    cv::imshow("Pitch result", pitch);
+                    cv::imshow("Image result", image);
+                    cv::waitKey(0);
+                    cv::destroyAllWindows();
+                } catch (const std::exception& ex) {
+                    logging::error("{}", ex.what());
+                }
+            })
+            .item("Reset", [this] {
+                _model.reset();
+                renderImages();
+            });
+
+        _gui.main_menu().button("Debug", [this] { _state.debugWindowFlag = not _state.debugWindowFlag; });
+    }
+
+    void renderWindows() {
+        if (_state.loadPopupFlag) {
+            _gui.popup("Load")
+                .input("path", &_model.imagePath())
+                .separator()
+                .button("Ok", vec2{ 120, 0 }, [this] {
```

Comment:
Factoring out the logic implementation of button. (e.g. named function instead of lambda)